examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py

```diff
@@ -0,0 +1,115 @@
+"""
+This component that captions images using a model from the Hugging Face hub.
+"""
+import io
+import itertools
+import logging
+import toolz
+
+from PIL import Image
+
+import dask
+import dask.dataframe as dd
+import pandas as pd
+
+from transformers import AutoProcessor, AutoModelForCausalLM
+import torch
+
+from fondant.component import FondantTransformComponent
+from fondant.logger import configure_logging
+
+configure_logging()
+logger = logging.getLogger(__name__)
+
+
+@dask.delayed
+def load(example):
+    bytes = io.BytesIO(example)
+    image = Image.open(bytes).convert("RGB")
+    return image
+
+
+@dask.delayed
+def transform(image, processor, device):
+    inputs = processor(images=image, return_tensors="pt").to(device)
+
+    return inputs
+
+
+@dask.delayed
+def collate(examples):
+    return torch.cat([ex["pixel_values"] for ex in examples])
+
+
+@dask.delayed
+def caption(batch, model, processor, max_new_tokens):
+    generated_ids = model.generate(pixel_values=batch, max_new_tokens=max_new_tokens)
+    generated_captions = processor.batch_decode(generated_ids, skip_special_tokens=True)
+
+    return generated_captions
+
+
+@dask.delayed
+def flatten(lst):
+    return pd.Series(itertools.chain(*lst))
+
+
+class CaptionImagesComponent(FondantTransformComponent):
+    """
+    Component that captions images using a model from the Hugging Face hub.
+    """
+
+    def transform(
+        self,
+        dataframe: dd.DataFrame,
+        model_id: str,
+        batch_size: int,
+        max_new_tokens: int,
+    ) -> dd.DataFrame:
+        """
+        Args:
+            dataframe: Dask dataframe
+            model_id: id of the model on the Hugging Face hub
+            batch_size: batch size to use
+            max_new_tokens: maximum token length of each caption
+
+        Returns:
+            Dask dataframe
+        """
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        processor = AutoProcessor.from_pretrained(model_id)
+        model = AutoModelForCausalLM.from_pretrained(model_id)
+
+        # load and transform the images
+        images = dataframe["images_data"]
+        loaded_images = [load(image) for image in images]
+        transformed_images = [
+            transform(image, processor, device) for image in loaded_images
```

Comment:
Also here, moving all the images to the device could be problematic (certainly if device='cuda'). This would give CUDA Out of memory